src/execution/operator/csv_scanner/state_machine/csv_state_machine_cache.cpp

```diff
@@ -70,26 +77,60 @@ void CSVStateMachineCache::Insert(const CSVStateMachineOptions &state_machine_op
 		}
 	}
 	// 2) Field Separator State
-	transition_array[delimiter][static_cast<uint8_t>(CSVState::DELIMITER)] = CSVState::DELIMITER;
-	transition_array[static_cast<uint8_t>('\n')][static_cast<uint8_t>(CSVState::DELIMITER)] =
-	    CSVState::RECORD_SEPARATOR;
-	if (new_line_id == NewLineIdentifier::CARRY_ON) {
-		transition_array[static_cast<uint8_t>('\r')][static_cast<uint8_t>(CSVState::DELIMITER)] =
-		    CSVState::CARRIAGE_RETURN;
-	} else {
-		transition_array[static_cast<uint8_t>('\r')][static_cast<uint8_t>(CSVState::DELIMITER)] =
-		    CSVState::RECORD_SEPARATOR;
-	}
 	transition_array[quote][static_cast<uint8_t>(CSVState::DELIMITER)] = CSVState::QUOTED;
-	if (delimiter != ' ') {
+	if (delimiter_first_byte != ' ') {
 		transition_array[' '][static_cast<uint8_t>(CSVState::DELIMITER)] = CSVState::EMPTY_SPACE;
 	}
-	if (comment != '\0') {
-		transition_array[comment][static_cast<uint8_t>(CSVState::DELIMITER)] = CSVState::COMMENT;
-	}
 
+	vector<uint8_t> delimiter_states {
+	    static_cast<uint8_t>(CSVState::DELIMITER), static_cast<uint8_t>(CSVState::DELIMITER_FIRST_BYTE),
+	    static_cast<uint8_t>(CSVState::DELIMITER_SECOND_BYTE), static_cast<uint8_t>(CSVState::DELIMITER_THIRD_BYTE)};
+
+	// These are the same transitions for all delimiter states
+	for (auto &state : delimiter_states) {
+		if (multi_byte_delimiter) {
+			transition_array[delimiter_first_byte][state] = CSVState::DELIMITER_FIRST_BYTE;
+		} else {
+			transition_array[delimiter_first_byte][state] = CSVState::DELIMITER;
+		}
+		transition_array[static_cast<uint8_t>('\n')][state] = CSVState::RECORD_SEPARATOR;
+		if (new_line_id == NewLineIdentifier::CARRY_ON) {
+			transition_array[static_cast<uint8_t>('\r')][state] = CSVState::CARRIAGE_RETURN;
+		} else {
+			transition_array[static_cast<uint8_t>('\r')][state] = CSVState::RECORD_SEPARATOR;
+		}
+		if (comment != '\0') {
+			transition_array[comment][static_cast<uint8_t>(CSVState::DELIMITER)] = CSVState::COMMENT;
+		}
+	}
+	// Deal other multi-byte delimiters
+	if (delimiter_value.size() == 2) {
+		transition_array[static_cast<uint8_t>(delimiter_value[1])]
+		                [static_cast<uint8_t>(CSVState::DELIMITER_FIRST_BYTE)] = CSVState::DELIMITER;
+	} else if (delimiter_value.size() == 3) {
+		transition_array[static_cast<uint8_t>(delimiter_value[1])]
+		                [static_cast<uint8_t>(CSVState::DELIMITER_FIRST_BYTE)] = CSVState::DELIMITER_SECOND_BYTE;
+		transition_array[static_cast<uint8_t>(delimiter_value[2])]
+		                [static_cast<uint8_t>(CSVState::DELIMITER_SECOND_BYTE)] = CSVState::DELIMITER;
+	} else if (delimiter_value.size() == 4) {
+		if (delimiter_value[0] == delimiter_value[2]) {
+			transition_array[static_cast<uint8_t>(delimiter_value[1])]
```

Comment:
Can we add a test where the first three bytes of the delimiter are the same, e.g. `AAAB`, with a csv file like:

```csv
Col1AAABCol2
AAAABC
```

Maybe also one with three (`AAB`)